File: finder4j-backtrack-examples/src/main/java/com/github/k0zka/finder4j/backtrack/examples/queens/QueensStep.java
package com.github.k0zka.finder4j.backtrack.examples.queens;

import com.github.k0zka.finder4j.backtrack.Step;

public class QueensStep implements Step<QueensState> {

	private static final long serialVersionUID = -555046442160555351L;

	public QueensStep(short queenNr, short pos) {
		super();
		this.queenNr = queenNr;
		this.pos = pos;
	}

	private final short queenNr;
	private final short pos;

	public QueensState take(QueensState state) {
		final Short[] queens = state.getQueens();
		queens[queenNr] = pos;
		return new QueensState(queens);
	}

	@Override
	public String toString() {
		return "{" + queenNr + "->" + pos + "}";
	}

	@Override
	public int hashCode() {
		return queenNr + pos;
	}

	@Override
	public boolean equals(Object obj) {
		return obj instanceof QueensStep && ((QueensStep)obj).queenNr == queenNr && ((QueensStep)obj).pos == pos;
	}

}

File: finder4j-backtrack-examples/src/main/kotlin/com/github/k0zka/finder4j/backtrack/examples/queens/QueensStep.kt
package com.github.k0zka.finder4j.backtrack.examples.queens

import com.github.k0zka.finder4j.backtrack.Step

class QueensStep(private val queenNr: Short, private val pos: Short) : Step<QueensState> {

	override fun take(state: QueensState): QueensState {
		val queens = state.queens
		queens[queenNr.toInt()] = pos
		return QueensState(queens)
	}

	override fun toString(): String {
		return "{$queenNr->$pos}"
	}

	override fun hashCode(): Int {
		return queenNr + pos
	}

	override fun equals(obj: Any?): Boolean {
		return obj is QueensStep && obj.queenNr == queenNr && obj.pos == pos
	}

	companion object {

		private val serialVersionUID = -555046442160555351L
	}

}
